File: app/src/main/java/com/felipecsl/elifut/services/LeagueRoundGenerator.java
package com.felipecsl.elifut.services;

import com.google.common.base.Preconditions;

import android.support.annotation.VisibleForTesting;
import android.util.Log;

import com.felipecsl.elifut.models.Club;
import com.felipecsl.elifut.models.LeagueRound;
import com.felipecsl.elifut.models.Match;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.felipecsl.elifut.util.CollectionUtils.shuffle;

public class LeagueRoundGenerator {
  private static final String TAG = "LeagueRoundGenerator";

  public LeagueRoundGenerator() {
  }

  @VisibleForTesting List<LeagueRound> generateRoundsDeterministic(List<? extends Club> clubs) {
    Preconditions.checkNotNull(clubs);
    Preconditions.checkArgument(clubs.size() > 1, "Need at least 2 clubs");

    List<? extends Club> clubsCopy = new ArrayList<>(clubs);

    if (clubsCopy.size() % 2 != 0) {
      clubsCopy.remove(clubsCopy.size() - 1);
      Log.d(TAG, "Got odd number of clubs, dropping the last one");
    }

    int totalClubs = clubsCopy.size();
    int totalRounds = (totalClubs - 1) * 2;
    int matchesPerRound = totalClubs / 2;
    LeagueRound[] rounds = new LeagueRound[totalRounds];

    for (int k = 0; k < rounds.length; k++) {
      rounds[k] = LeagueRound.create(k + 1);
    }
    Map<Integer, Club> clubMap = new HashMap<>();
    for (int i = 1; i <= totalClubs; i++) {
      clubMap.put(i, clubsCopy.get(i - 1));
    }

    for (int round = 0; round < totalRounds; round++) {
      for (int match = 0; match < matchesPerRound; match++) {
        int home = (round + match) % (totalClubs - 1);
        int away = (totalClubs - 1 - match + round) % (totalClubs - 1);

        // Last team stays in the same place while the others
        // rotate around it.
        if (match == 0) {
          away = totalClubs - 1;
        }

        Club clubHome;
        Club clubAway;
        if (round % 2 == 0) {
          clubHome = clubMap.get(home + 1);
          clubAway = clubMap.get(away + 1);
        } else {
          clubHome = clubMap.get(away + 1);
          clubAway = clubMap.get(home + 1);
        }

        Match newMatch = Match.create(clubHome, clubAway);
        rounds[round].addMatch(newMatch);
      }
    }

    return Arrays.asList(rounds);
  }

  /**
   * Generates a random list of league rounds from the provided list of clubs.
   */
  public List<LeagueRound> generateRounds(List<? extends Club> clubs) {
    return generateRoundsDeterministic(shuffle(clubs));
  }
}

File: app/src/main/kotlin/com/elifut/services/LeagueRoundGenerator.kt
package com.elifut.services

import android.support.annotation.VisibleForTesting
import android.util.Log
import android.util.SparseArray
import com.elifut.models.Club
import com.elifut.models.LeagueRound
import com.elifut.models.Match
import com.elifut.util.shuffle
import com.google.common.base.Preconditions
import java.util.*

open class LeagueRoundGenerator {
  @VisibleForTesting fun generateRoundsDeterministic(clubs: List<Club>): List<LeagueRound> {
    Preconditions.checkNotNull(clubs)
    Preconditions.checkArgument(clubs.size > 1, "Need at least 2 clubs")

    val clubsCopy = ArrayList(clubs)

    if (clubsCopy.size % 2 != 0) {
      clubsCopy.removeAt(clubsCopy.size - 1)
      Log.d(TAG, "Got odd number of clubs, dropping the last one")
    }

    val totalClubs = clubsCopy.size
    val totalRounds = (totalClubs - 1) * 2
    val matchesPerRound = totalClubs / 2
    val rounds = arrayOfNulls<LeagueRound>(totalRounds)

    for (k in rounds.indices) {
      rounds[k] = LeagueRound.create(k + 1)
    }
    val clubMap = mutableMapOf<Int, Club>()
    for (i in 1..totalClubs) {
      clubMap.put(i, clubsCopy[i - 1])
    }

    for (round in 0..totalRounds - 1) {
      for (match in 0..matchesPerRound - 1) {
        val home = (round + match) % (totalClubs - 1)
        var away = (totalClubs - 1 - match + round) % (totalClubs - 1)

        // Last team stays in the same place while the others
        // rotate around it.
        if (match == 0) {
          away = totalClubs - 1
        }

        val clubHome: Club
        val clubAway: Club
        if (round % 2 == 0) {
          clubHome = clubMap[home + 1]!!
          clubAway = clubMap[away + 1]!!
        } else {
          clubHome = clubMap[away + 1]!!
          clubAway = clubMap[home + 1]!!
        }

        val newMatch = Match.create(clubHome, clubAway)
        rounds[round]!!.addMatch(newMatch)
      }
    }

    return Arrays.asList<LeagueRound>(*rounds)
  }

  /** Generates a random list of league rounds from the provided list of clubs.  */
  open fun generateRounds(clubs: List<Club>): List<LeagueRound> {
    // Make a defensive copy since we need to mutate it
    return generateRoundsDeterministic(ArrayList(clubs).shuffle())
  }

  companion object {
    private val TAG = "LeagueRoundGenerator"
  }
}
